src/main/java/com/scalar/db/storage/dynamo/SelectStatementHandler.java

```diff
@@ -0,0 +1,267 @@
+    QueryRequest.Builder builder = QueryRequest.builder().tableName(dynamoOperation.getTableName());
+
+    setConditions(builder, scan);
+
+    if (!scan.getOrderings().isEmpty()) {
+      scan.getOrderings()
+          .forEach(
+              o -> {
+                if (dynamoOperation.isSortKey(o.getName())
+                    && o.getOrder() == Scan.Ordering.Order.DESC) {
+                  builder.scanIndexForward(false);
+                }
+              });
+    }
+
+    if (scan.getLimit() > 0) {
+      builder.limit(scan.getLimit());
+    }
+
+    if (!scan.getProjections().isEmpty()) {
+      builder.projectionExpression(String.join(",", scan.getProjections()));
+    }
+
+    if (scan.getConsistency() != Consistency.EVENTUAL) {
+      builder.consistentRead(true);
+    }
+
+    return client.query(builder.build());
+  }
+
+  private void setConditions(QueryRequest.Builder builder, Scan scan) {
+    List<String> conditions = new ArrayList<>();
+
+    conditions.add(getPartitionKeyCondition());
+
+    boolean isRangeEnabled = setRangeCondition(scan, conditions);
+    if (!isRangeEnabled) {
+      setStartCondition(scan, conditions);
+      setEndCondition(scan, conditions);
+    }
+    String keyConditions = String.join(" AND ", conditions);
+
+    Map<String, AttributeValue> bindMap = getPartitionKeyBindMap(scan);
+    if (isRangeEnabled) {
+      bindMap.putAll(getRangeBindMap(scan));
+    }
+    bindMap.putAll(getStartBindMap(scan, isRangeEnabled));
+    bindMap.putAll(getEndBindMap(scan, isRangeEnabled));
+
+    builder.keyConditionExpression(keyConditions).expressionAttributeValues(bindMap);
+  }
+
+  private String getPartitionKeyCondition() {
+    return DynamoOperation.PARTITION_KEY + " = " + DynamoOperation.PARTITION_KEY_ALIAS;
+  }
+
+  private Map<String, AttributeValue> getPartitionKeyBindMap(Scan scan) {
+    Map<String, AttributeValue> bindMap = new HashMap<>();
+    DynamoOperation dynamoOperation = new DynamoOperation(scan, metadataManager);
+    String partitionKey = dynamoOperation.getConcatenatedPartitionKey();
+
+    bindMap.put(
+        DynamoOperation.PARTITION_KEY_ALIAS, AttributeValue.builder().s(partitionKey).build());
+
+    return bindMap;
+  }
+
+  private boolean setRangeCondition(Scan scan, List<String> conditions) {
+    if (!scan.getStartClusteringKey().isPresent() || !scan.getEndClusteringKey().isPresent()) {
+      return false;
+    }
+
+    List<Value> start = scan.getStartClusteringKey().get().get();
+    List<Value> end = scan.getEndClusteringKey().get().get();
+    String startKeyName = start.get(start.size() - 1).getName();
+    String endKeyName = end.get(end.size() - 1).getName();
+
+    DynamoOperation dynamoOperation = new DynamoOperation(scan, metadataManager);
+
+    if (startKeyName.equals(endKeyName)) {
+      conditions.add(startKeyName + DynamoOperation.RANGE_CONDITION);
+      if (!scan.getStartInclusive() || !scan.getEndInclusive()) {
+        throw new IllegalArgumentException(
```

Comment:
Similar to the clustering key limitation, it probably needs to exclude entries in the adapter side not to break the abstraction.

BTW, `com.amazonaws.services.dynamodbv2` is an old library?
https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ScanJavaDocumentAPI.html
In this, Scan seems to support exclusive for start, but not for end...

Reply:
Thank you.
I'm not sure what `exclusiveStartKey` does. It seems that it doesn't suite this usage.

> To fetch the next page, you execute the scan method again by providing the primary key value of the last item in the previous page so that the scan method can return the next set of items. You provide this information in the request by using the withExclusiveStartKey method.`

> The primary key of the first item that this operation will evaluate. Use the value that was returned for LastEvaluatedKey in the previous operation.